data_model/primitives/src/conststr.rs

```diff
@@ -0,0 +1,650 @@
+//! Const-string related implementation and structs.
+#[cfg(not(feature = "std"))]
+use alloc::{borrow::ToOwned, boxed::Box, str::from_utf8_unchecked, string::String};
+use core::{
+    cmp::{Eq, Ord, Ordering, PartialEq, PartialOrd},
+    convert::TryFrom,
+    fmt,
+    hash::{Hash, Hasher},
+    mem::{size_of, ManuallyDrop},
+    ops::Deref,
+    slice::{from_raw_parts, from_raw_parts_mut},
+};
+#[cfg(feature = "std")]
+use std::str::from_utf8_unchecked;
+
+use derive_more::{DebugCustom, Display};
+use iroha_schema::{IntoSchema, MetaMap};
+use parity_scale_codec::{WrapperTypeDecode, WrapperTypeEncode};
+use serde::{
+    de::{Deserialize, Deserializer, Error, Visitor},
+    ser::{Serialize, Serializer},
+};
+
+const MAX_INLINED_STRING_LEN: usize = 2 * size_of::<usize>() - 1;
+
+/// Immutable inlinable string.
+/// Strings shorter than 15/7/3 bytes (in 64/32/16-bit architecture) are inlined.
+/// Union represents const-string variants: inlined or boxed.
+/// Distinction between variants are achieved by tagging most significant bit of field `len`:
+/// - for inlined variant MSB of `len` is always equal to 1, it's enforced by `InlinedString` constructor;
+/// - for boxed variant MSB of `len` is always equal to 0, it's enforced by the fact
+/// that `Box` and `Vec` never allocate more than`isize::MAX bytes`.
+/// For little-endian 64bit architecture memory layout of `ConstStringData` is following:
+///
+/// ```text
+/// +---------+-------+---------+----------+----------------+
+/// | Bits    | 0..63 | 64..118 | 119..126 | 127            |
+/// +---------+-------+---------+----------+----------------+
+/// | Inlined | payload         | len      | tag (always 1) |
+/// +---------+-------+---------+----------+----------------+
+/// | Box     | ptr   | len                | tag (always 0) |
+/// +---------+-------+--------------------+----------------+
+/// ```
+#[derive(Display, DebugCustom)]
+#[display(fmt = "{}", "&**self")]
+#[debug(fmt = "{:?}", "&**self")]
+#[repr(C)]
+pub union ConstString {
+    inlined: InlinedString,
+    boxed: ManuallyDrop<BoxedString>,
+}
+
+impl ConstString {
+    /// Returns the length of this [`ConstString`], in bytes.
```

Comment:
In general, function docstrings should be written in imperative mood. I.e. 
```suggestion
    /// Return the length of [`Self`], in bytes.
```

Reply:
Using `Self` instead of `ConstString` is a good convention, it also saves you from having to check if you renamed the doc-comments.